src/Raven.Server/Documents/Replication/ReplicationTopologyDestinationExplorerResponder.cs

```diff
@@ -0,0 +1,168 @@
+                    tcp.DocumentDatabase,
+                    header.AlreadyVisited,
+                    TimeSpan.FromMilliseconds(header.Timeout),
+                    replicationDocument.Destinations))
+                {
+                    //if true, all adjacent nodes are already visited, so no need to continue
+                    //(filtering is done in a ctor of ReplicationTopologyDiscoverer, using header.AlreadyVisited)
+                    if (topologyDiscoverer.DiscovererCount == 0)
+                    {
+                        var topology = GetFullTopologyWithLocalNodes(localDbId, localTopology);
+                        WriteDiscoveryResponse(tcp, context, topology, null , TopologyDiscoveryResponse.Status.Ok);
+                        return;
+                    }
+
+                    try
+                    {
+                        var discoveryTask = topologyDiscoverer.DiscoverTopologyAsync();
+                        discoveryTask.Wait(tcp.DocumentDatabase.DatabaseShutdown);
+
+                        var topology = GetFullTopologyWithLocalNodes(localDbId, localTopology);
+                        foreach (var nodeInfo in discoveryTask.Result.NodesByDbId)
+                            topology.NodesByDbId[nodeInfo.Key] = nodeInfo.Value;
+
+                        WriteDiscoveryResponse(tcp, context, topology, null , TopologyDiscoveryResponse.Status.Ok);
+                    }
+                    catch (Exception e)
+                    {
+                        var topology = GetFullTopologyWithLocalNodes(localDbId, localTopology);
+                        WriteDiscoveryResponse(
+                            tcp, 
+                            context,
+                            topology,
+                            e.ToString(),
+                            TopologyDiscoveryResponse.Status.Error);
+                    }
+                }
+            }
+        }
+
+        private static FullTopologyInfo GetFullTopologyWithLocalNodes(string localDbId, NodeTopologyInfo localTopology)
+        {
+            var topology = new FullTopologyInfo(localDbId);
+            topology.NodesByDbId.Add(localDbId, localTopology);
+            return topology;
+        }
+
+        private void UpdateAlreadyVisitedWithLocalDestinations(
+            string originDbId,
+            TcpConnectionOptions tcp,
+            Dictionary<string, List<string>> alreadyVisited)
+        {
+            List<string> visitedDbIds;
+            if (!alreadyVisited.TryGetValue(originDbId, out visitedDbIds))
+            {
+                visitedDbIds = new List<string>();
+                alreadyVisited.Add(originDbId,visitedDbIds);
+            }
+
+            var localDbId = tcp.DocumentDatabase.DbId.ToString();
+            if (!visitedDbIds.Contains(localDbId))
+                visitedDbIds.Add(localDbId);
+
+            foreach (var outgoing in tcp.DocumentDatabase.DocumentReplicationLoader.OutgoingHandlers)
```

Comment:
You are also still using the `OutgoingHandlers`, instead of using `Destinations`